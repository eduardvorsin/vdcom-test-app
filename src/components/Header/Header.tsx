import {
  AppBar,
  Box,
  IconButton,
  SxProps,
  Theme,
  styled,
} from '@mui/material';
import MenuIcon from '@mui/icons-material/Menu';
import React from 'react';
import ProfileImg from '../../assets/images/profile.jpg?webp';
import Profile from '../Profile/Profile';
import Search from '../Search/Search';
import Logo from '../Logo/Logo';

type HeaderProps = {
  sx?: SxProps<Theme>,
  onMenuButtonClick: React.MouseEventHandler<HTMLButtonElement>,
}

const StyledAppBar = styled(AppBar)(() => ({
  position: 'absolute',
  justifyContent: 'center',
}));

const StyledToolbar = styled(Box)(({ theme }) => ({
  maxWidth: '100%',
  margin: '0',
  display: 'grid',
  justifyContent: 'start',
  alignItems: 'center',
  padding: '5px 15px',
  gap: '10px 20px',
  gridTemplateColumns: 'min-content min-content 1fr',
  gridTemplateRows: '1fr 1fr',
  [theme.breakpoints.up('md')]: {
    gridTemplateRows: '1fr',
    gridTemplateColumns: '1fr 215px',
    padding: '5px 24px',
    justifyContent: 'space-between',
  },
}));

const BurgerMenu = styled(IconButton)(({ theme }) => ({
  fontSize: '0px',
  color: '#000000',
  [theme.breakpoints.up('md')]: {
    display: 'none',
  },
}));

const Header: React.FC<HeaderProps> = ({
  sx,
  onMenuButtonClick,
}) => (
  <StyledAppBar
    sx={sx}
  >
    <StyledToolbar
    >
      <BurgerMenu
        onClick={onMenuButtonClick}
      >
        <MenuIcon />
        open navigation menu
      </BurgerMenu>

      <Logo
        sx={{
          display: { sx: 'block', md: 'none' },
        }}

      />

      <Search
        value={''}
        onChange={() => { }}
        sx={{
          maxWidth: { lg: '552px', md: '350px', xs: '250px' },
          gridRow: { xs: '2/3', md: '1/2' },
          gridColumn: { xs: '1/-1', md: 'span 1' },
        }}
      />

      <Profile
        name='Mr. Director'
        role='Managing Director'
        src={ProfileImg}
        sx={{
          marginLeft: { xs: 'auto', md: '0' },
        }}
      />
    </StyledToolbar>
  </StyledAppBar>
);

export default Header;
