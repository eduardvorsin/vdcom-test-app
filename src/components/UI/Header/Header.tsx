import {
  AppBar,
  Box,
  IconButton,
  SxProps,
  Theme,
  styled,
} from '@mui/material';
import MenuIcon from '@mui/icons-material/Menu';
import React, { useState } from 'react';
import ProfileImg from '../../../assets/images/profile.jpg?webp';
import Profile from '../Profile/Profile';
import Search from '../../Search/Search';
import Logo from '../Logo/Logo';

type HeaderProps = {
  sx?: SxProps<Theme>,
  onMenuButtonClick: React.MouseEventHandler<HTMLButtonElement>,
}

const StyledAppBar = styled(AppBar)(() => ({
  position: 'absolute',
  justifyContent: 'center',
  zIndex: '5',
}));

const AppBarInner = styled(Box)(({ theme }) => ({
  maxWidth: '100%',
  margin: '0',
  display: 'grid',
  justifyContent: 'start',
  alignItems: 'center',
  padding: '5px 15px',
  gap: '10px 20px',
  gridTemplateColumns: 'min-content min-content 1fr',
  gridTemplateRows: '1fr 1fr',
  [theme.breakpoints.up('md')]: {
    gridTemplateRows: '1fr',
    gridTemplateColumns: '1fr 215px',
    padding: '5px 24px',
    justifyContent: 'space-between',
  },
}));

const BurgerMenu = styled(IconButton)(({ theme }) => ({
  fontSize: '0px',
  color: '#000000',
  [theme.breakpoints.up('md')]: {
    display: 'none',
  },
}));

const Header: React.FC<HeaderProps> = ({
  sx,
  onMenuButtonClick,
}) => {
  const [searchValue, setSearchValue] = useState<string>('');

  const searchChangeHandler: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    setSearchValue(e.target.value);
  };

  return (
    <StyledAppBar
      sx={sx}
    >
      <AppBarInner
      >
        <BurgerMenu
          onClick={onMenuButtonClick}
        >
          <MenuIcon />
          open navigation menu
        </BurgerMenu>

        <Logo
          sx={{
            display: { sx: 'block', md: 'none' },
          }}

        />

        <Search
          id='main-search'
          value={searchValue}
          onChange={searchChangeHandler}
          sx={{
            maxWidth: { lg: '552px', md: '350px', sm: '250px', xs: '100%' },
            gridRow: { xs: '2/3', md: '1/2' },
            gridColumn: { xs: '1/-1', md: 'span 1' },
          }}
        />

        <Profile
          name='Mr. Director'
          role='Managing Director'
          src={ProfileImg}
          alt='Mr. Director avatar'
        />
      </AppBarInner>
    </StyledAppBar>
  );
};

export default Header;
