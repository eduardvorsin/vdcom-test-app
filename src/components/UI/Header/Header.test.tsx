import React from 'react';

import { render, screen, waitFor } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';

import userEvent from '@testing-library/user-event';
import Header from './Header';

const mockFn = jest.fn<void, [e: React.MouseEvent<HTMLButtonElement>]>();
const routes = [{
  path: '/',
  element: <Header onMenuButtonClick={mockFn} />,
}];
const router = createMemoryRouter(routes, { initialEntries: ['/'] });

describe('Header tests', () => {
  it('renders correctly', () => {
    render(<RouterProvider router={router} />);

    expect(screen.getByRole<HTMLElement>('banner')).toBeInTheDocument();
    expect(screen.getByRole<HTMLButtonElement>('button', { name: /open navigation menu/i })).toBeInTheDocument();
    expect(screen.getByRole<HTMLDivElement>('search')).toBeInTheDocument();
  });

  it('should call the mock function when clicking on the "open navigation menu" button', async () => {
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    user.click(screen.getByRole<HTMLButtonElement>('button', { name: /open navigation menu/i }));

    await waitFor(() => {
      expect(mockFn).toHaveBeenCalledTimes(1);
    });
  });

  it('is a basic snapshot', () => {
    render(<RouterProvider router={router} />);

    expect(screen.getByRole<HTMLElement>('banner')).toMatchSnapshot();
  });
});

describe('Header integration tests', () => {
  it('should be the displayed value if it has been entered', async () => {
    const user = userEvent.setup();
    render(<RouterProvider router={router} />);

    user.type(screen.getByRole<HTMLInputElement>('textbox'), 'contacts');

    await waitFor(() => {
      expect(screen.getByRole<HTMLInputElement>('textbox')).toHaveValue('contacts');
    });
  });
});
