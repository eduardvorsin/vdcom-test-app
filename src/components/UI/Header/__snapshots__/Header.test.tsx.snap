// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tests is a basic snapshot 1`] = `
<header
  class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-4efeke-MuiPaper-root-MuiAppBar-root"
  data-testid="header"
>
  <div
    class="MuiBox-root css-sdx3s9"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-ubps8a-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="MenuIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
      open navigation menu
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <a
      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-112rk81-MuiTypography-root-MuiLink-root"
      data-testid="logo-link"
      href="/contacts"
    >
      <img
        alt="logo"
        class="css-12pywbb"
        src="assets"
      />
    </a>
    <form
      class="css-1pho46g"
      data-testid="search-form"
      novalidate=""
    >
      <div
        class="MuiBox-root css-19sk4h4"
        role="search"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-13pwdle-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </button>
        <label
          class="css-inngth"
          for="main-search"
        >
          Search about documents
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-15tv2wk-MuiInputBase-root-MuiInput-root"
          required=""
        >
          <input
            aria-describedby="main-search-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="main-search"
            name="search"
            placeholder="Search by Name..."
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
    <div
      class="MuiBox-root css-ro3nde"
      data-testid="profile"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular css-quq592-MuiAvatar-root"
      >
        <img
          alt="Mr. Director avatar"
          class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="assets"
        />
      </div>
      <div
        class=" MuiBox-root css-yi99sy"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-14tr2nh-MuiTypography-root"
        >
          Mr. Director
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-j1f0n7-MuiTypography-root"
        >
          Managing Director
        </p>
      </div>
    </div>
  </div>
</header>
`;
